### ISSUE TITLE
buf `breaking --against .git#…` can only be used with local branch names

### ISSUE THREAD

**Issue by powerman** [bufbuild/buf#339]
I believe everything about this issue was already said in #303, but as was requested in #338 here is reproducible example:
```sh
$ mkdir buf-non-local-branch
$ cd buf-non-local-branch
$ mkdir first
$ cd first 
$ git init
Initialized empty Git repository in /home/powerman/tmp/buf-non-local-branch/first/.git/
$ echo 'version: v1beta1' > buf.yaml
$ touch service.proto
$ git add .
$ git commit -m one
[master (root-commit) c3dc449] one
 2 files changed, 1 insertion(+)
 create mode 100644 buf.yaml
 create mode 100644 service.proto
$ cd ..
$ git clone first second
Cloning into 'second'...
done.
$ cd second
$ git checkout -b example
Switched to a new branch 'example'
$ git commit --allow-empty -m two
[example a04d5c1] two
$ buf breaking --against '.git#branch=master'
$ buf breaking --against '.git#ref=origin/master'
$ git branch -d master 
Deleted branch master (was c3dc449).
$ buf breaking --against '.git#branch=master'
Failure: could not clone file:///home/powerman/tmp/buf-non-local-branch/second/.git: exit status 128
Cloning into ''...
warning: Could not find remote branch master to clone.
fatal: Remote branch master not found in upstream origin
.
$ buf breaking --against '.git#ref=origin/master'
Failure: could not clone file:///home/powerman/tmp/buf-non-local-branch/second/.git: exit status 1
error: pathspec 'origin/master' did not match any file(s) known to git
.
$ 
```

**Comment by bufdev** [comment]
In this example, you deleted the `master` branch - what is your expectation here?

**Comment by powerman** [comment]
My expectation is what on CI we often have only current PR's branch and no other local branches, including master.

**Comment by powerman** [comment]
I mean, in this case we usually still have origin/master, so while `.git#branch=master` should break in this case I'd like to have `.git#ref=origin/master` working.

**Comment by powerman** [comment]
Moreover, I've just noticed another case (probably it's for another issue, but just in case you'll decide to fix something near this functionality I'll mention it here):
1. Set `master` and `origin/master` to commit with proto **version A** (usual case).
2. Set branch `feature` to commit with proto **version B** with incompatible change (also usual case).
3. At this time `buf breaking` will complain both for `.git#branch=master` and `.git#ref=origin/master` (so everything is correct so far).
4. Now, move the local `master` branch to the `feature` branch:
```
$ git checkout master
$ git reset --hard feature
```
At this moment you'll see `buf breaking` doesn't report any errors. And while it's reasonable behaviour for `.git#branch=master` (because `master` branch is also on proto **version B** now), it looks like a bug for `.git#ref=origin/master` case (because `origin/master` is still on proto **version A**). Switching from the local `master` to the `feature` branch (which is on same commit now) doesn't makes any difference.

**Comment by bufdev** [comment]
We've updated the git reference cloning change the way references and branches are treated. Previously, when specifying `.git#branch=origin/master`, Buf would `git clone` the repository at `$PWD/.git` and then checkout the ref `origin/master`, which would be confusing as `origin/master` referred to the `master` branch (after the clone `origin` would be `$PWD/.git`).

In the new behavior, instead of doing a clone we init a new repository and add `$PWD/.git` as a remote, after which we fetch the ref user supplied ref from the remote. This causes the ref to be interpreted in the context of the origin (`$PWD/.git`), and results in fetching the ref that makes more sense.

This will go out in 0.47.0.

This should resolve this issue, let us know otherwise!

**Comment by bufdev** [comment]
This is released in 0.47.0.
